examples/vault-iam-auth/user-data-auth-client.sh

```diff
@@ -0,0 +1,104 @@
+#!/bin/bash
+# This script is meant to be run in the User Data of each EC2 Instance while it's booting. The script uses the
+# run-consul script to configure and start Consul in client mode. Note that this script assumes it's running in an AMI
+# built from the Packer template in examples/vault-consul-ami/vault-consul.json.
+
+set -e
+
+# Send the log output from this script to user-data.log, syslog, and the console
+# From: https://alestic.com/2010/12/ec2-user-data-output/
+exec > >(tee /var/log/user-data.log|logger -t user-data -s 2>/dev/console) 2>&1
+
+# These variables are passed in via Terraform template interpolation
+/opt/consul/bin/run-consul --client --cluster-tag-key "${consul_cluster_tag_key}" --cluster-tag-value "${consul_cluster_tag_value}"
+
+
+function retry {
+  local readonly cmd=$1
+  local readonly description=$2
+
+  for i in $(seq 1 30); do
+    echo "$description"
+
+    # The boolean operations with the exit status are there to temporarily circumvent the "set -e" at the
+    # beginning of this script which exits the script immediatelly for error status while not losing the exit status code
+    output=$(eval "$cmd") && exit_status=0 || exit_status=$?
```

Comment:
Are you using `output` anywhere? If not, any reason to use a subshell and store the output like that?

Reply:
When retrying the login the output is used to get retrieve the client token

Reply:
I don't see any reference to `output` anywhere in this function? I see a reference further down in the code, outside of the function though... Is that intentional? Are you relying on the side effect of the `retry` function happening to set a variable called `output`? If so, that's quite confusing and non-obvious. I'd recommend restructuring to avoid that sort of leap.

Reply:
Yes, I'm relying on the side-effect since bash functions only return integers. 
I could `echo` the output and then `$()` but then we loose the helpful echos with retry messages, so I opted for this compromise...  Do you know a better way where we can keep both?

Reply:
Yup! You can log to `stderr` and use `stdout` for return values. See https://github.com/gruntwork-io/bash-commons/blob/master/modules/bash-commons/src/log.sh.

Reply:
I could send a reference too... that might be the best compromise

Reply:
Ah! let me check that

Reply:
worked like a charm